src/ORM/LazyEagerLoader.php

```diff
@@ -0,0 +1,163 @@
+
+        $query = $source
+            ->find()
+            ->select((array)$primaryKey)
+            ->where(function ($exp, $q) use ($primaryKey, $keys, $source) {
+                if (is_array($primaryKey) && count($primaryKey) === 1) {
+                    $primaryKey = current($primaryKey);
+                }
+
+                if (is_string($primaryKey)) {
+                    return $exp->in($source->aliasField($primaryKey), $keys->toList());
+                }
+
+                $types = array_intersect_key($q->defaultTypes(), array_flip($primaryKey));
+                $primaryKey = array_map([$source, 'aliasField'], $primaryKey);
+                return new TupleComparison($primaryKey, $keys->toList(), $types, 'IN');
+            })
+            ->contain($contain);
+
+
+        foreach ($query->eagerLoader()->attachableAssociations($source) as $loadable) {
+            $config = $loadable->config();
+            $config['includeFields'] = true;
+            $loadable->config($config);
+        }
+
+        return $query;
+    }
+
+    /**
+     * Returns a map of property names where the association results should be injected
+     * in the top level entities.
+     *
+     * @param Cake\ORM\Table $source The table having the top level associations
+     * @param array $associations The name of the top level associations
+     * @return array
+     */
+    protected function _getPropertyMap($source, $associations)
+    {
+        $map = [];
+        foreach ($associations as $assoc) {
+            $map[$assoc] = $source->associations()->get($assoc)->property();
+        }
+        return $map;
+    }
+
+    /**
+     * Injects the results of the eager loader query into the original list of
+     * entities.
+     *
+     * @param array|Traversable $objects The original list of entities
+     * @param Cake\Collection\CollectionInterface $results The loaded results
+     * @param array $associations The top level associations that were loaded
+     * @param Cake\ORM\Table $source The table where the entities came from
+     * @return array
+     */
+    protected function _injectResults($objects, $results, $associations, $source)
+    {
+        $injected = [];
+        $properties = $this->_getPropertyMap($source, $associations);
+        $primaryKey = (array)$source->primaryKey();
+        $results = $results
+            ->indexBy(function ($e) use ($primaryKey) {
+                return implode(';', $e->extract($primaryKey));
+            })
+            ->toArray();
+
+        foreach ($objects as $k => $object) {
+            $key = implode(';', $object->extract($primaryKey));
+            if (!isset($results[$key])) {
+                $injected[$k] = $object;
+                continue;
+            }
+
+            $loaded = $results[$key];
+            foreach ($associations as $assoc) {
+                $property = $properties[$assoc];
+                $object->set($property, $loaded->get($property), ['useSetters' => false]);
```

Comment:
This is clever, I was wondering how you were going to reconstruct the contained entities correctly.